scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,370 @@
+/* Copyright (c) Andrea Righi <andrea.righi@canonical.com> */
+/*
+ * scx_rustlite: simple vtime-based scheduler written in Rust
+ *
+ *  The main goal of this scheduler is be an "easy to read" template that can
+ *  be used to quickly test more complex scheduling policies. For this reason
+ *  this scheduler is mostly focused on simplicity and code readability.
+ *
+ * The scheduler is made of a BPF part that implements the basic sched-ext
+ * functionalities and a user-space counterpart, written in Rust, that
+ * implement the scheduling policy itself.
+ *
+ * The default policy used by the user-space part is a based on virtual runtime
+ * (vtime):
+ *
+ * - each task receives the same time slice of execution
+ *
+ * - the actual execution time (adjusted based on the task's static priority)
+ *   determines the vtime
+ *
+ * - tasks are then dispatched from the lowest to the highest vtime
+ *
+ * The vtime is evaluated in the BPF part, tasks' then all the information are
+ * sent to the user-space part and stored in a red-black tree (indexed by
+ * vtime); then the user-space part sends back to the BPF part the list of
+ * tasks (ordered by their vtime, that will be dispatched using the order
+ * determined by the user-space.
+ *
+ * This software may be used and distributed according to the terms of the
+ * GNU General Public License version 2.
+ */
+#include <scx/common.bpf.h>
+#include "intf.h"
+
+char _license[] SEC("license") = "GPL";
+
+/*
+ * Exit info (passed to the user-space counterpart).
+ */
+int exit_kind = SCX_EXIT_NONE;
+char exit_msg[SCX_EXIT_MSG_LEN];
+
+/*
+ * Scheduler attributes and statistics.
+ */
+u32 usersched_pid; /* User-space scheduler PID */
+
+static bool usersched_needed; /* Used to wake-up the user-space scheduler */
+const volatile bool switch_partial; /* Switch all tasks or SCHED_EXT tasks */
+const volatile u64 slice_ns = SCX_SLICE_DFL; /* Base time slice duration */
+
+u64 nr_enqueues, nr_user_dispatches, nr_kernel_dispatches; /* Statistics */
+
+/*
+ * Maximum amount of tasks enqueued/dispatched between kernel and user-space at
+ * a certain time.
+ *
+ * The enqueued/dispatched queues are used in a producer/consumer way between
+ * the BPF part and the user-space part.
+ */
+#define MAX_ENQUEUED_TASKS 1024
+
+/*
+ * Tasks enqueued to the user-space for scheduling.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_QUEUE);
+	__uint(key, 0);
+	__type(value, struct task_ctx);
+	__uint(max_entries, MAX_ENQUEUED_TASKS);
+} enqueued SEC(".maps");
+
+/*
+ * Tasks enqueued by the user-space for dispatching.
```

Comment:
I was actually considering to rename those queue maps to somehting like `to_userspace` and `from_userspace`. Maybe it's a bit naive, but it would really help to understand that they're used for kernel/user-space communication.